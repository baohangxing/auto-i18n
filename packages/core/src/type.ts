export interface I18nCallRule {
  /** caller of functionName
       *  @default ''
      */
  caller: string
  /** function name of `t`
       *  @default 't'
      */
  functionName: string

  definedDeclaration: string
  /**
       * import i18n t Declaration
       */
  importDeclaration: string
}

export type FileExtension = 'js' | 'ts' | 'cjs' | 'mjs' | 'jsx' | 'tsx' | 'vue'

export interface AutoConfig {
  /**
      * local Locales-Json directory Paths, all user json file should found in this paths
      *
      * @example [`/Users/XXX/GM/src/lang/locales`]
      */
  localesJsonDirs: string[] | string

  /**
       * locales used, the locale will not trans if not in `locales`
       *
       * @example [`ja-jp`, `zh-cn`]
       */
  locales: string[]

  /**
       * the base locale in your project, make sure it is in `locales` and all words has configured in locale-Json file
       *
       * @default `zh-cn`
       */
  baseLocale: string

  /**
       *
       * @param locale
       * @returns
       */
  untransSymbol: (locale: string) => string

  /**
 * translate word rule when not translated
 *
 * @param word
 * @param locale the word current locale, your baseLocale
 * @param toLocale to translate locale like `ja-jp`
 * @returns
 */
  transLacaleWord?: (word: string, locale: string, toLocale: string) => Promise<string>

  /**
      * files to auto trans, using fast-glob rule
      *
      * see deteils: https://github.com/mrmlnc/fast-glob
      *
      * @example ["src/*.{vue}", "view/**"]
      */
  includes: string[]

  /**
       * exclude files to auto trans, using fast-glob rule
       *
       * see deteils: https://github.com/mrmlnc/fast-glob
       *
       * @example  ['**\/node_modules/**\/*']
       */
  exclude: string[]

  /**
       * output filed directory path,
       * the default output filed directory path is your project path
       *
       * @default `./ `
       */
  outputFileDir: string

  i18nCallRules: Record<FileExtension, I18nCallRule>
}

export interface TransCommandOption {
  filePath: string
  generateNewFile: boolean
  newFileName: string
}

export interface LangJson {
  name: string
  path: string
}
