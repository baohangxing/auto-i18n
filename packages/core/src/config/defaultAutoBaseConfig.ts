import type { AutoBaseConfig, I18nCallRule } from '../types'

const getCommonI18nCallRule = (): I18nCallRule => {
  return {
    transCaller: '',
    transIdentifier: 't',
    variableDeclaration: 'const { t } = useI18n();',
    importDeclaration: 'import { useI18n } from "vue-i18n";',
  }
}

const defaultAutoBaseConfig: AutoBaseConfig = {
  localesJsonDirs: ['**/locales/**.json', '!test/**'],
  locales: [],
  baseLocale: '',

  transInterpolationsMode: 'NamedInterpolationMode',
  i18nCallRules: {
    js: getCommonI18nCallRule(),
    ts: getCommonI18nCallRule(),
    cjs: getCommonI18nCallRule(),
    mjs: getCommonI18nCallRule(),
    jsx: getCommonI18nCallRule(),
    tsx: getCommonI18nCallRule(),
    vue: getCommonI18nCallRule(),
  },
}

export default defaultAutoBaseConfig
