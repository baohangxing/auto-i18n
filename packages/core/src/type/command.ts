export interface TransCommandOption {
  filePath: string
  generateNewFile: boolean
  newFileName: string
}
