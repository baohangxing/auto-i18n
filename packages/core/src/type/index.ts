export * from './command'
export * from './autoConfig'
