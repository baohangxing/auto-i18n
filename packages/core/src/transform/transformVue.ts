import type {
  SFCDescriptor,
  SFCScriptBlock,
  SFCStyleBlock,
  SFCTemplateBlock,
} from '@vue/compiler-sfc'
import { parse } from '@vue/compiler-sfc'
import * as htmlparser2 from 'htmlparser2'
import mustache from 'mustache'
import ejs from 'ejs'
import presetTypescript from '@babel/preset-typescript'
import type { I18nCallRule } from '../types'
import { includeChinese } from '../utils/help'
import log from '../utils/log'
import { IGNORE_REMARK } from '../config/constants'
import transformJs from './transformJs'
import { initParse } from './parse'
import Collector from './collector'
import { escapeQuotes, getCallExpression, getCallExpressionPrefix, trimValue } from './tools'

type Handler = (source: string, rule: I18nCallRule, replace: boolean) => string

const COMMENT_TYPE = '!'

const parseJsSyntax = (source: string, rule: I18nCallRule): string => {
  // vue属性有可能是{xx:xx}这种对象形式，直接解析会报错，需要特殊处理。
  // 先处理成temp = {xx:xx} 让babel解析，解析完再还原成{xx:xx}
  let isObjectStruct = false
  if (source.startsWith('{') && source.endsWith('}')) {
    isObjectStruct = true
    source = `___temp = ${source}`
  }
  const { code } = transformJs(source, {
    rule: {
      ...rule,
      variableDeclaration: '',
      importDeclaration: '',
    },
    parse: initParse([[presetTypescript, { isTSX: true, allExtensions: true }]]),
  })

  let stylizedCode = code

  if (isObjectStruct)
    stylizedCode = stylizedCode.replace('___temp = ', '')

  stylizedCode = stylizedCode.replace(/;$/gm, '')
  return stylizedCode.endsWith('\n') ? stylizedCode.slice(0, stylizedCode.length - 1) : stylizedCode
}

const removeQuotes = (value: string): string => {
  if (['"', '\''].includes(value.charAt(0)) && ['"', '\''].includes(value.charAt(value.length - 1)))
    value = value.substring(1, value.length - 1)

  return value
}

const handleTemplate = (code: string, rule: I18nCallRule, replace = true): string => {
  let htmlString = ''

  const getReplaceValue = (value: string): string => {
    value = escapeQuotes(value)
    return getCallExpression(rule, Collector.getKey(value))
  }

  let shouldIgnore = false
  const parser = new htmlparser2.Parser(
    {
      onopentag(tagName, attributes) {
        let attrs = ''
        if (shouldIgnore) {
          for (const key in attributes) {
            const attrValue = attributes[key]
            attrs += ` ${key}="${attrValue}" `
          }
          htmlString += `<${tagName} ${attrs}>`
          return
        }

        for (const key in attributes) {
          const attrValue = attributes[key]
          const isVueDirective = key.startsWith(':') || key.startsWith('@') || key.startsWith('v-')
          if (includeChinese(attrValue) && isVueDirective) {
            const source = parseJsSyntax(attrValue, rule)
            // 处理属性类似于:xx="'xx'"，这种属性值不是js表达式的情况
            // attrValue === source即属性值不是js表达式
            // attrValue.startsWith是为了排除:xx="t('中文')"的情况
            if (attrValue === source && !attrValue.startsWith(getCallExpressionPrefix(rule))) {
              Collector.add(removeQuotes(attrValue))
              if (replace)
                attrs += ` ${key}="${getReplaceValue(removeQuotes(attrValue))}"`
            }
            else {
              attrs += ` ${key}="${source}"`
            }
          }
          else if (includeChinese(attrValue) && !isVueDirective) {
            Collector.add(attrValue)
            if (replace)
              attrs += ` :${key}="${getReplaceValue(attrValue)}"`
          }
          else if (attrValue === '') {
            attrs += ` ${key}`
          }
          else {
            attrs += ` ${key}="${attrValue}"`
          }
        }
        htmlString += `<${tagName} ${attrs}>`
      },

      ontext(text) {
        if (shouldIgnore) {
          htmlString += text
          return
        }
        let str = ''
        const tokens = mustache.parse(text)
        for (const token of tokens) {
          const type = token[0]
          let value = token[1]

          if (includeChinese(value)) {
            value = trimValue(value)
            if (type === 'text') {
              Collector.add(value)
              if (replace)
                str += `{{${getReplaceValue(value)}}}`
            }
            else if (type === 'name') {
              const source = parseJsSyntax(value, rule)
              str += `{{${source}}}`
            }
          }
          else {
            if (type === 'text') {
              str += value
            }
            else if (type === 'name') {
              str += `{{${value}}}`
            }
            else if (type === COMMENT_TYPE) {
              // 形如{{!xxxx}}这种形式，在mustache里属于注释语法
              str += `{{!${value}}}`
            }
          }
        }

        htmlString += str
      },

      onclosetag(tagName, isImplied) {
        shouldIgnore = false
        // 如果是自闭合标签
        if (isImplied) {
          htmlString = `${htmlString.slice(0, htmlString.length - 2)}/>`
          return
        }
        htmlString += `</${tagName}>`
      },

      oncomment(comment) {
        if (comment.includes(IGNORE_REMARK))
          shouldIgnore = true

        htmlString += `<!--${comment}-->`
      },
    },
    {
      lowerCaseTags: false,
      recognizeSelfClosing: true,
      lowerCaseAttributeNames: false,
      // xmlMode: true,
    },
  )

  parser.write(code)
  parser.end()
  return htmlString
}

const handleScript = (source: string, rule: I18nCallRule, replace: boolean): string => {
  const { code } = transformJs(source, {
    rule,
    isJsInVue: true,
    parse: initParse([[presetTypescript, { isTSX: true, allExtensions: true }]]),
  },
  replace)
  return `\n${code}\n`
}

const mergeCode = (templateCode: string, scriptCode: string, stylesCode: string): string => {
  return `${templateCode}\n${scriptCode}\n${stylesCode}`
}

const getWrapperTemplate = (sfcBlock: SFCTemplateBlock | SFCScriptBlock | SFCStyleBlock): string => {
  const { type, lang, attrs } = sfcBlock
  let template = `<${type}`

  if (lang)
    template += ` lang="${lang}"`

  if ((sfcBlock as SFCScriptBlock).setup)
    template += ' setup'

  if ((sfcBlock as SFCStyleBlock).scoped)
    template += ' scoped'

  for (const attr in attrs) {
    if (!['lang', 'scoped', 'setup'].includes(attr))
      template += ` ${attr}="${attrs[attr]}"`
  }
  template += `><%- code %></${type}>`
  return template
}

const generateSource = (
  sfcBlock: SFCTemplateBlock | SFCScriptBlock,
  handler: Handler,
  rule: I18nCallRule,
  replace: boolean,
): string => {
  const wrapperTemplate = getWrapperTemplate(sfcBlock)
  const source = handler(sfcBlock.content, rule, replace)
  return ejs.render(wrapperTemplate, {
    code: source,
  })
}

const removeSnippet = (
  source: string,
  sfcBlock: SFCTemplateBlock | SFCScriptBlock | SFCStyleBlock | null,
): string => {
  return sfcBlock ? source.replace(sfcBlock.content, '') : source
}

/**
 * 提取文件头注释 //TODO fix in 1.0.0
 * @param descriptor
 * @returns
 */
const getFileComment = (descriptor: SFCDescriptor): string => {
  const { template, script, scriptSetup, styles } = descriptor
  let source = descriptor.source
  source = removeSnippet(source, template)
  source = removeSnippet(source, script)
  source = removeSnippet(source, scriptSetup)
  if (styles) {
    for (const style of styles)
      source = removeSnippet(source, style)
  }
  const result = source.match(/<!--[\s\S]*?-->/m)
  return result ? result[0] : ''
}

const transformVue = (
  code: string,
  rule: I18nCallRule,
  replace = true,
): {
  code: string
} => {
  const { descriptor, errors } = parse(code)
  if (errors.length > 0) {
    log.error('parse vue error', errors[0].toString())
    return {
      code,
    }
  }

  const { template, script, scriptSetup, styles } = descriptor
  let templateCode = ''
  let scriptCode = ''
  let stylesCode = ''

  const fileComment = getFileComment(descriptor)

  if (template)
    templateCode = generateSource(template, handleTemplate, rule, replace)

  if (script)
    scriptCode = generateSource(script, handleScript, rule, replace)

  if (scriptSetup)
    scriptCode = generateSource(scriptSetup, handleScript, rule, replace)

  if (styles) {
    for (const style of styles) {
      const wrapperTemplate = getWrapperTemplate(style)
      const source = style.content
      stylesCode
        += `${ejs.render(wrapperTemplate, {
          code: source,
        })}\n`
    }
  }

  code = mergeCode(templateCode, scriptCode, stylesCode)
  if (fileComment)
    code = `${fileComment}\n${code}`

  return {
    code,
  }
}

export default transformVue
