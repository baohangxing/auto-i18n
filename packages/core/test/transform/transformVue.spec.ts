import { describe, expect, it } from 'vitest'
import type { Collector } from '../../src/types'
import type { TransformVueOptions } from '../../src/transform/transformVue'
import transformVue from '../../src/transform/transformVue'
import defaultAutoBaseConfig from '../../src/config/defaultAutoBaseConfig'

const genTransOptions = (collector: Collector): TransformVueOptions => {
  return {
    rule: {
      transCaller: '',
      transIdentifier: 't',
      variableDeclaration: 'const t = useI18n()',
      importDeclaration: 'import { useI18n } from \'vue-i18n\'',
    },
    replace: true,
    collector,
    autoConfig: defaultAutoBaseConfig,
  }
}

describe('#transformVue', () => {
  it('should transform vue', async () => {
    const words: string[] = []
    const reslut = transformVue(
    `<template>
      <div>这是{{title}}</div>
    </template>
    <script setup>
      let title = "标题"
    </script>
    `,
    genTransOptions(
      {
        add(str) {
          words.push(str)
        },
        getKey(str: string) {
          return `Key of ${str}`
        },
      }),
    )
    expect(words).toEqual(['这是', '标题'])
    expect(reslut.code).toMatchSnapshot()
  })

  it('should transform attributes in vue template', async () => {
    const words: string[] = []
    const reslut = transformVue(
    `<template>
      <div>这{{title}}</div>
      <comp v-if="'吃'"> 这是 {{ "什么东西啊" }}</comp>
      <comp @click="()=>{a=t('我啊')}" :title="'好呢'"/>
      <!--auto-i18n-ignore-->
      <comp>忽略</comp>
    </template>
    <script setup>
      let title = "标题"
    </script>
    `,
    genTransOptions(
      {
        add(str) {
          words.push(str)
        },
        getKey(str: string) {
          return `Key of ${str}`
        },
      }),
    )
    expect(words.sort()).toEqual([
      '吃',
      '好呢',
      '标题',
      '这是',
      '这',
    ].sort())
    expect(reslut.code).toMatchSnapshot()
  })

  it('should transform attributes in vue template', async () => {
    const words: string[] = []
    const reslut = transformVue(
    `<template>
    <div class="home">
      <img alt="Vue logo" src="../assets/logo.png"/>
      <HelloWorld :msg="'你好' + '世界'"/>
      {{ hi }}
    </div>
  </template>
  
  <script lang="ts" setup>
  import HelloWorld from '@/components/HelloWorld.vue'
  
  const hi = '你好a'
  
  </script>
    `,
    genTransOptions(
      {
        add(str) {
          words.push(str)
        },
        getKey(str: string) {
          return `Key of ${str}`
        },
      }),
    )
    expect(words.sort()).toEqual(['世界', '你好', '你好a'].sort())
    expect(reslut.code).toMatchSnapshot()
  })
})
