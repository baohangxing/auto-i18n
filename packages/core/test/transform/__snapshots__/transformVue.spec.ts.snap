// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#transformVue > should transform attributes in vue template 1`] = `
"<template>
      <div>{{t('Key of 这')}}{{title}}</div>
      <comp v-if=\\"t('Key of 吃')\\">{{t('Key of 这是')}}{{\\"什么东西啊\\"}}</comp>
      <comp @click=\\"() => {
  a = t('我啊')
}\\" :title=\\"t('Key of 好呢')\\"/>
      <!--auto-i18n-ignore-->
      <comp>忽略</comp>
    </template>

<script setup>
import { useI18n } from 'vue-i18n';
const t = useI18n();
let title = t('Key of 标题');
</script>

"
`;

exports[`#transformVue > should transform attributes in vue template 2`] = `
"<template>
    <div class=\\"home\\">
      <img alt=\\"Vue logo\\" src=\\"../assets/logo.png\\"/>
      <HelloWorld :msg=\\"t('Key of 你好') + t('Key of 世界')\\"/>
      {{hi}}
    </div>
  </template>

<script lang=\\"ts\\" setup>
import HelloWorld from '@/components/HelloWorld.vue';
import { useI18n } from 'vue-i18n';
const t = useI18n();
const hi = t('Key of 你好a');
</script>

"
`;

exports[`#transformVue > should transform vue 1`] = `
"<template>
      <div>{{t('Key of 这是')}}{{title}}</div>
    </template>

<script setup>
import { useI18n } from 'vue-i18n';
const t = useI18n();
let title = t('Key of 标题');
</script>

"
`;
